### PR TITLE
AAD Authentication Options

### LINKED ISSUE #321: AAD Authentication Improvements
Currently it's possible to enable App Service Auth ("easy auth") for a FlowEHR app. This is good but does not give us the options we need for other types of apps, specifically:
- RBAC
- Single Page Apps - since there's no client side access to the token

## Description

As a FlowEHR App Developer
I want to be given the ID and identifier of an AAD app that is pre-configured and ready to use for my app. I want to be able to assign users into roles and determine the role membership when the user logs in to my app.
So that I can offer more advanced and fit-for-purpose apps to my organisation.

## Acceptance criteria

- [x] `apps.*.config` extended to allow a FlowEHR admin to specify `easy_auth`, `app_roles`, and `single_page_application` settings on an app-by-app basis
- [x] Relevant secrets to be fed into GH environment and App Service config to be used by the hosted app at build and run time

### PR DESCRIPTION
**Pull request by damoodamoo**
# REsolves #321 

## What is being addressed

AAD App creation extended to cover:
- Ability to create an app and _not_ assign it as an easy auth app, which allows Single Page App scenarios
- Add `single_page_application` block to config
- Add `app_roles`
- Collapse testing slot app and main slot app into one, so they share the same AAD app.

## Has the CIS document been consulted / updated?

N/A

### CODE REVIEW

File: apps/app/aad_app/locals.tf [hunk 1]
```diff
@@ -12,9 +12,6 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
-module "aad_app" {
```
Comment (damoodamoo) [inline apps/app/aad_app/locals.tf:15]: not sure why GitHub thinks this was even in `locals.tf`, it wasn't

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/6014613685 (with refid `69cd7490`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/322#issuecomment-1697721315) from @damoodamoo)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/6014800077 (with refid `69cd7490`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/322#issuecomment-1697752376) from @damoodamoo)

**Review by jjgriff93 — APPROVED** [review]
Looks good!

File: apps/app/github-environment.tf [hunk 2]
```diff
@@ -121,6 +121,32 @@ resource "github_actions_environment_secret" "acr_image_name" {
   ]
 }
 
+resource "github_actions_environment_secret" "aad_auth_client_id" {
+  for_each        = local.branches_and_envs
+  repository      = local.github_repository_name
+  environment     = each.value
+  secret_name     = "AAD_AUTH_CLIENT_ID"
+  plaintext_value = try(module.aad_app[0].aad_application_id, "")
```
[the remark this answers is not in the document]
Comment (damoodamoo) [reply apps/app/github-environment.tf:129]: The module is conditional on optional config, so the try is a null check - I can try without it if you don't think I need it?

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/6016206537 (with refid `69cd7490`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/322#issuecomment-1697943816) from @damoodamoo)

File: apps/app/aad_app/main.tf [hunk 3]
```diff
@@ -35,23 +39,58 @@ resource "azuread_application" "webapp_auth" {
...
-    redirect_uris = ["https://${var.auth_webapp_name}.azurewebsites.net/.auth/login/aad/callback"]
+  dynamic "web" {
+    for_each = var.auth_settings.easy_auth ? [1] : []
+
+    content {
+      homepage_url = "https://${var.webapp_name}.azurewebsites.net"
+      redirect_uris = compact(concat([
+        "https://${var.webapp_name}.azurewebsites.net/.auth/login/aad/callback",
+        length(var.testing_slot_webapp_name) > 0 ? "https://${var.testing_slot_webapp_name}.azurewebsites.net/.auth/login/aad/callback" : null
+      ]))
```
Comment (stefpiatek) [inline apps/app/aad_app/main.tf:61]: does this `concat` need to be here?

**Review by stefpiatek — APPROVED** [review]
Looks nice 🎉

**Pull request merged**